src/peer_select.cc

```diff
@@ -256,154 +390,131 @@ PeerSelector::CheckAlwaysDirectDone(allow_t answer, void *data)
-            fs->code = ORIGINAL_DST; // fs->code is DIRECT. This fixes the display.
-            handlePath(p, *fs);
+    for (auto it = aclPeersCache.begin(); it != aclPeersCache.end(); ++it) {
+        if (p == it->first.valid()) {
+            answer = it->second;
+            return true;
         }
-
-        // clear the used fs and continue
-        servers = fs->next;
-        delete fs;
-        resolveSelected();
-        return;
     }
+    return false;
+}
 
-    // convert the list of FwdServer destinations into destinations IP addresses
-    if (fs && wantsMoreDestinations()) {
-        // send the next one off for DNS lookup.
-        const char *host = fs->_peer.valid() ? fs->_peer->host : request->url.host();
-        debugs(44, 2, "Find IP destination for: " << url() << "' via " << host);
-        Dns::nbgethostbyname(host, this);
-        return;
-    }
+static void
+checkLastPeerAccessWrapper(allow_t answer, void *data)
+{
+    PeerSelector *selector = static_cast<PeerSelector *>(data);
+    selector->checkLastPeerAccess(answer);
+}
 
-    // Bug 3605: clear any extra listed FwdServer destinations, when the options exceeds max_foward_tries.
-    // due to the allocation method of fs, we must deallocate each manually.
-    // TODO: use a std::list so we can get the size and abort adding whenever the selection loops reach Config.forward_max_tries
-    if (fs) {
-        assert(fs == servers);
-        while (fs) {
-            servers = fs->next;
-            delete fs;
-            fs = servers;
-        }
-    }
+void
+PeerSelector::checkLastPeerAccess(allow_t answer)
+{
 
-    // done with DNS lookups. pass back to caller
+    if (currentServer->_peer.valid())
+        aclPeersCache.push_back(std::pair<CbcPointer<CachePeer>, allow_t>(currentServer->_peer, answer));
 
-    debugs(44, 2, id << " found all " << foundPaths << " destinations for " << url());
-    debugs(44, 2, "  always_direct = " << always_direct);
-    debugs(44, 2, "   never_direct = " << never_direct);
-    debugs(44, 2, "       timedout = " << ping.timedout);
+    if (!answer.allowed() || !currentServer->_peer.valid()) {
+        currentServer = currentServer->next;
+        selectMore();
+        return;
+    }
 
-    ping.stop = current_time;
-    request->hier.ping = ping; // final result
+    if (currentServer->groupId)
+        groupSelect(currentServer);
 
-    if (lastError && foundPaths) {
-        // nobody cares about errors if we found destinations despite them
-        debugs(44, 3, "forgetting the last error");
-        delete lastError;
-        lastError = nullptr;
-    }
+    updateSelectedPeer(currentServer->_peer.get(), currentServer->code);
 
-    if (const auto initiator = interestedInitiator())
-        initiator->noteDestinationsEnd(lastError);
-    lastError = nullptr; // initiator owns the ErrorState object now
-    delete this;
+    if (callback_ != nullptr) {
```

Comment:
difficult to be null but it can be canceled. 
I removed the check. The PeerSelector::callback asserts if callback_ is missing.
But I added a check for selectionAborted at the top of  PeerSelector::checkLastPeerAccess.